[GStreamer] Fix crash on devices where video sink is auto-plugged

eocanha opened the pull request:
#### 869bf0a64f0f7cb1c606447e259a738ed908a187
<pre>
[GStreamer] Fix crash on devices where video sink is auto-plugged
<a href="https://bugs.webkit.org/show_bug.cgi?id=254674">https://bugs.webkit.org/show_bug.cgi?id=254674</a>

Reviewed by Xabier Rodriguez-Calvar.

Some (downstream) platforms don&apos;t have any specific videosink enforced and use autoplugging instead.
This means that createVideoSink() will return nullptr and the sink will be created automatically
(autoplugged) when the time comes. However, the lambda[1] that sets max-lateness at createVideoSink()
triggers an assert if the sink is null at that point.

Original author: Eugene Mutavchi &lt;Ievgen_Mutavchi@comcast.com&gt;

See: <a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1059">https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1059</a>

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::createVideoSink): Removed assert and just return if m_videoSink is null.

Canonical link: <a href="https://commits.webkit.org/265066@main">https://commits.webkit.org/265066@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ac638b677a7f845b2c6c2ab5f18de27c9342530

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1119 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1149 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1188 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1787 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/987 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1103 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1199 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1229 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1182 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1127 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1040 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1040 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1662 "Built successfully") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1038 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1023 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1079 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1067 "7 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2114 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1020 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/993 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1066 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1047 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2328 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1006 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1087 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3939,8 +3939,7 @@ GstElement* MediaPlayerPrivateGStreamer::createVideoSink()
 
     // Ensure the sink has the max-lateness property set.
     auto exit = makeScopeExit([this] {
-        ASSERT(m_videoSink);
```
Why remove this but not the one before returning the method result?

eocanha replied:
Because this patch focuses only on the holepunch case, used on Nexus Broadcom platforms. In that case, [createHolePunchVideoSink() is called and it returns null to explicitly defer to autovideosink](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L3907), so [createVideoSink() returns early](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L3973) and the final assert before returning the method result that you point out isn't called. However, [the makeScopeExit lambda](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L3940) is still called automatically before returning.

philn replied:
I see... :( Maybe we can simplify this.... How about calling `createVideoSink()` only in `!USE(WPEWEBKIT_PLATFORM_BCM_NEXUS)` cases?

eocanha replied:
That would mean splitting the holepunch setup functionality in two, so it simplifies on one place to create complexity and unneeded care of implementation details on the other.

philn replied:
Sorry I don't see what you mean.

eocanha replied:
I mean that a decision that is now taken in [createHolePunchVideoSink()](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L3907) (because it's very related to holepunch) has now to be taken "standalone" (without any holepunch context) in [createGSTPlaybin()](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L2871), without any mention to holepunch, which is the place where taking the decision to defer to autoplugging comes from.

philn replied:
I see no mention of holepunch in this code, beside its current location...
```
#if USE(WPEWEBKIT_PLATFORM_BCM_NEXUS)
    // Nexus boxes use autovideosink.
    return nullptr;
#endif
```

philn replied:
I give up, this spaghetti is beyond control.

eocanha committed:
[GStreamer] Fix crash on devices where video sink is auto-plugged
https://bugs.webkit.org/show_bug.cgi?id=254674

Reviewed by Xabier Rodriguez-Calvar.

Some (downstream) platforms don't have any specific videosink enforced and use autoplugging instead.
This means that createVideoSink() will return nullptr and the sink will be created automatically
(autoplugged) when the time comes. However, the lambda[1] that sets max-lateness at createVideoSink()
triggers an assert if the sink is null at that point.

Original author: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>

See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1059

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::createVideoSink): Removed assert and just return if m_videoSink is null.

Canonical link: https://commits.webkit.org/265066@main
